### PR TITLE
WIP:  Use new gpu runner for array-api Actions workflow

### PR DESCRIPTION
**Pull request by mataton**
## Add GPU/Array API CI workflow

Adds a new GitHub Actions workflow (`array-api.yml`) that runs array-API-decorated tests on a T4 GPU runner against PyTorch, JAX, and CuPy with CUDA. Tests are filtered via a new `array_api` pytest marker so only relevant tests run.

### Changes

- **New workflow** (`array-api.yml`): runs weekly (Mondays 06:00 UTC), on manual dispatch, or when a PR is labeled `gpu-ci`. Verifies each backend sees the GPU at install time, then runs `pytest -m array_api`.
- **`@array_backends` decorator** now applies a `pytest.mark.array_api` marker so CI can filter on it. Marker registered in `pyproject.toml`.
- **Skip guards tightened** in `test_array.py`: replaced "is the module importable" checks with usability probes that actually exercise the backend. Fixes spurious errors when a library installs but can't allocate (e.g., CuPy without a GPU).
- **`_normalize_device` replaced** with `_device_specs_match` and `_on_device`. The old function blurred two concerns; the new split distinguishes "do these spec strings refer to the same hardware" from "is this array actually on that hardware."
- **`assert_type_preserved`** now uses `_on_device` to inspect real array device placement instead of string-comparing device reprs.
- **Developer docs**: new `devdoc/array_api.rst` covering how to write array-API library code and tests, run them locally, and use the `gpu-ci` label.

### Notes

- Dask was considered but excluded. Its support in scikit-bio is dev-only and not fully built out; adding CI commitment now would be premature.
- Workflow is intentionally not a required check — GPU runners can be flaky, and most PRs don't touch array code. Reviewers apply the `gpu-ci` label when validation is needed.
- Only 16 tests are currently decorated (in `composition` and `ordination`). Coverage will grow as more of the package adopts the array API.

Please complete the following checklist:

* [x] I have read the [contribution guidelines](https://scikit.bio/contribute.html).

* [ ] I have documented all public-facing changes in the [changelog](https://github.com/scikit-bio/scikit-bio/blob/main/CHANGELOG.md).

* [x] **This pull request includes code, documentation, or other content derived from external source(s).** If this is the case, ensure the external source's license is compatible with scikit-bio's [license](https://github.com/scikit-bio/scikit-bio/blob/main/LICENSE.txt). Include the license in the `licenses` directory and add a comment in the code giving proper attribution. Ensure any other requirements set forth by the license and/or author are satisfied.

  - **It is your responsibility to disclose** code, documentation, or other content derived from external source(s). If you have questions about whether something can be included in the project or how to give proper attribution, include those questions in your pull request and a reviewer will assist you.

* [ ] This pull request does not include code, documentation, or other content derived from external source(s).

Note: [This document](https://scikit.bio/devdoc/review.html) may also be helpful to see some of the things code reviewers will be verifying when reviewing your pull request.

### CODE REVIEW

**Comment by qiyunzhu** [comment]
Looks like we need to patch the unit test for `_on_device` with some dummy namespaces in order to bring up the coverage, or mark it to skip coverage calculation. @mataton

**Comment by mataton** [comment]
@qiyunzhu Still need to add a similar workflow to test things on CPU, but other than that this should be ready for review.

File: skbio/util/tests/test_array.py [hunk 1]
```diff
@@ -25,15 +25,18 @@
...
-dask = get_package("dask", raise_error=False)
-if dask is not None:
-    da = get_package("dask.array")
+da = get_package("dask.array", raise_error=False)
+
+CUPY_OK = _is_usable(cp, lambda: cp.arange(1))
+TORCH_OK = _is_usable(torch, lambda: torch.arange(1))
+JAX_OK = _is_usable(jax, lambda: jax.numpy.arange(1))
+DASK_OK = _is_usable(da, lambda: da.arange(1, chunks=1).compute())
+XARRAY_OK = _is_usable(xr, lambda: xr.DataArray([1]))
```
Comment (qiyunzhu) [inline skbio/util/tests/test_array.py:37]: Do we have official Xarray support? If not, I am not sure if we should include Xarray tests. I also think that Xarray is more similar to dataframes instead of arrays. I have an intension to add Xarray to the "table-like" support list. But Xarray can handle more than two dimensions, and we haven't explicited dealt with this logic.

File: .github/workflows/array-api.yml [hunk 2]
```diff
@@ -1,37 +1,85 @@
...
+      - name: Check CUDA driver
+        run: nvidia-smi
+
       - name: Install scikit-bio + backend
         run: |
-          pip install -e .
-          pip install ${{ matrix.backend }}
+          python -m ensurepip
+          pip install -v .
+          case "${{ matrix.backend }}" in
```
Comment (qiyunzhu) [inline .github/workflows/array-api.yml:47]: Looks like a specific CUDA version is pinned. It will be good to add a comment on how to update CUDA version in the future.

File: .github/workflows/array-api.yml [hunk 3]
```diff
@@ -1,37 +1,85 @@
...
+  latest_python: "3.13"   # torch has no 3.14 wheels yet
 
 jobs:
   test:
+    if: |
+      github.event_name != 'pull_request' ||
+      contains(github.event.pull_request.labels.*.name, 'gpu-ci')
     name: Array API (${{ matrix.backend }})
-    runs-on: ${{ matrix.runs-on }}
+    runs-on: gpu-t4-4-core
```
Comment (qiyunzhu) [inline .github/workflows/array-api.yml:21]: Can add a comment here that `gpu-t4-4-core` is a specific runner configured in the scikit-bio organization.

File: .github/workflows/array-api.yml [hunk 4]
```diff
@@ -1,37 +1,85 @@
...
 
 env:
-  latest_python: "3.14"
+  latest_python: "3.13"   # torch has no 3.14 wheels yet
 
 jobs:
   test:
+    if: |
+      github.event_name != 'pull_request' ||
+      contains(github.event.pull_request.labels.*.name, 'gpu-ci')
```
Comment (qiyunzhu) [inline .github/workflows/array-api.yml:18]: Can add a comment here that the workflow will be triggered when the `gpu-ci` tag is added to the PR.

File: web/devdoc/array_api.rst [hunk 5]
```diff
@@ -0,0 +1,217 @@
+Array API and GPU support
+=========================
+
+Scikit-bio is gradually adopting the `Python array API standard
+<https://data-apis.org/array-api/latest/>`_ to allow functions to work
+across multiple array libraries (NumPy, PyTorch, JAX, CuPy) and run on
```
Comment (qiyunzhu) [inline web/devdoc/array_api.rst]: Shall we add "Dask" to this list?

File: skbio/util/tests/test_array.py (hunk 1 above)
Comment (mataton) [reply skbio/util/tests/test_array.py:37]: The Xarray tests were from the original `test_array.py`, I didn't add them so I didn't want to remove them. But if you think we can remove them then I think that's okay. The only test it's used in is for `ingest_array` specifically.

**Comment by qiyunzhu** [comment]
Matt and I went through the code and discussed offline. We agreed that it can be merged.
